pkg/frontend/querymiddleware/astmapper/instant_splitting.go

```diff
@@ -0,0 +1,542 @@
+			Args:     expr.Args,
+			PosRange: expr.PosRange,
+		},
+		RHS: &parser.Call{
+			Func:     parser.Functions[string(countOverTime)],
+			Args:     expr.Args,
+			PosRange: expr.PosRange,
+		},
+	}
+
+	// If avg_over_time is wrapped by a vector aggregator,
+	// the embedded query cannot be sent downstream
+	if i.embeddedAggregatorExpr != nil {
+		i.embeddedAggregatorExpr = nil
+	}
+
+	return i.MapNode(avgOverTimeExpr, stats)
+}
+
+// mapCallRate maps a rate function to expression increase / rangeInterval
+func (i *instantSplitter) mapCallRate(expr *parser.Call, rangeInterval time.Duration) (mapped parser.Node, finished bool, err error) {
+	increaseExpr := &parser.Call{
+		Func:     parser.Functions[string(increase)],
+		Args:     expr.Args,
+		PosRange: expr.PosRange,
+	}
+
+	// If rate is wrapped by a vector aggregator,
+	// the embedded query also needs to be updated to use increase
+	if i.embeddedAggregatorExpr != nil {
+		updatedExpr := updateEmbeddedExpr(i.embeddedAggregatorExpr, increaseExpr)
+		if updatedExpr == nil {
+			i.embeddedAggregatorExpr = nil
+		}
+	}
+
+	mappedNode, finished, err := i.mapCallByRangeInterval(increaseExpr, rangeInterval, parser.SUM)
+	if err != nil {
+		return nil, false, err
+	}
+	if !finished {
+		return mapped, false, nil
+	}
+
+	mappedExpr, ok := mappedNode.(parser.Expr)
+	if !ok {
+		return nil, false, fmt.Errorf("unable to map expr '%s'", expr)
+	}
+
+	return &parser.BinaryExpr{
+		Op:  parser.DIV,
+		LHS: mappedExpr,
+		RHS: &parser.NumberLiteral{Val: rangeInterval.Seconds()},
+	}, true, nil
+}
+
+func (i *instantSplitter) mapCallByRangeInterval(expr *parser.Call, rangeInterval time.Duration, op parser.ItemType) (mapped parser.Node, finished bool, err error) {
+	// Default grouping is 'without' for concatenating the embedded queries
+	var grouping []string
+	groupingWithout := true
+	if i.embeddedAggregatorExpr != nil {
+		grouping = append(grouping, i.embeddedAggregatorExpr.Grouping...)
+		groupingWithout = i.embeddedAggregatorExpr.Without
+	}
+
+	embeddedExpr, finished, err := i.splitAndSquashCall(expr, rangeInterval)
+	if err != nil {
+		return nil, false, err
+	}
+	if !finished {
+		return expr, false, nil
+	}
+
+	return &parser.AggregateExpr{
+		Op:       op,
+		Expr:     embeddedExpr,
+		Param:    nil,
+		Grouping: grouping,
+		Without:  groupingWithout,
+		PosRange: parser.PositionRange{},
+	}, true, nil
+}
+
+// expr is the range vector aggregator expression
+// If the outer expression is a vector aggregator, r.embeddedAggregatorExpr will contain the expression
+// In this case, the vector aggregator should be downstream to the embedded queries in order to limit
+// the label cardinality of the parallel queries
+func (i *instantSplitter) splitAndSquashCall(expr *parser.Call, rangeInterval time.Duration) (mapped parser.Expr, finished bool, err error) {
+	splitCount := int(math.Ceil(float64(rangeInterval) / float64(i.splitByInterval)))
+	if splitCount <= 0 {
```

Comment:
Should be `<= 1`? Don't we want to keep the input `expr` as if `rangeInterval < splitByInterval`?

Reply:
Yes, you are right. Great catch 👀